File: Design-Patterns-Using-JavaScript/Visitor/VisitorShopping.js
class Element {
    price() {
        throw new Error("price method must be implemented");
    }

    accept(visitor) {
        throw new Error("accept method must be implemented");
    }
}

class Book extends Element {
    constructor(price, isbn) {
        super();
        this.priceValue = price;
        this.isbn = isbn;
    }

    price() {
        return this.priceValue;
    }

    accept(visitor) {
        return visitor.visitBook(this);
    }
}

class Fruit extends Element {
    constructor(price, quantity, type) {
        super();
        this.priceValue = price;
        this.quantity = quantity;
        this.type = type;
    }

    price() {
        return this.priceValue;
    }

    accept(visitor) {
        return visitor.visitFruit(this) * this.quantity;
    }
}

class Visitor {
    visitBook(book) {
        throw new Error("visitBook method must be implemented");
    }

    visitFruit(fruit) {
        throw new Error("visitFruit method must be implemented");
    }
}

class SundayDiscount extends Visitor {
    visitBook(book) {
        return book.price() - 50;
    }

    visitFruit(fruit) {
        return fruit.price() - 5;
    }
}

class SaleDiscount extends Visitor {
    visitBook(book) {
        return book.price() / 2;
    }

    visitFruit(fruit) {
        return fruit.price() / 2;
    }
}

class ShoppingCart {
    constructor() {
        this.list = [];
        this.visitor = null;
    }

    add(element) {
        this.list.push(element);
    }

    setDiscountVisitor(discount) {
        this.visitor = discount;
    }

    accept() {
        let cost = 0;
        for (const element of this.list) {
            cost += element.accept(this.visitor);
        }
        console.log("Total cost: " + cost);
    }
}

// Client code
const cart = new ShoppingCart();
cart.add(new Fruit(100, 10, "Apple"));
cart.add(new Book(100, 12345));

cart.setDiscountVisitor(new SundayDiscount());
cart.accept();

cart.setDiscountVisitor(new SaleDiscount());
cart.accept();

/*
Total cost: 1000
Total cost: 550
*/

File: Design-Patterns-Using-Ruby/Visitor/VisitorShopping.rb
class Element
    def accept(visitor)
        raise NotImplementedError, 'Subclasses must implement the accept method'
    end

    def price
        raise NotImplementedError, 'Subclasses must implement the price method'
    end
end

class Book < Element
    attr_reader :price, :isbn

    def initialize(price, isbn)
        @price = price
        @isbn = isbn
    end

    def price
        @price
    end

    def accept(visitor)
        visitor.visit_book(self)
    end
end

class Fruit < Element
    attr_reader :price, :quantity, :type

    def initialize(price, quantity, type)
        @price = price
        @quantity = quantity
        @type = type
    end

    def price
        @price
    end

    def accept(visitor)
        visitor.visit_fruit(self) * quantity
    end
end

class Visitor
    def visit_book(book)
        raise NotImplementedError, 'Subclasses must implement the visit_book method'
    end

    def visit_fruit(fruit)
        raise NotImplementedError, 'Subclasses must implement the visit_fruit method'
    end
end

class SundayDiscount < Visitor
    def visit_book(book)
        book.price - 50
    end

    def visit_fruit(fruit)
        fruit.price - 5
    end
end

class SaleDiscount < Visitor
    def visit_book(book)
        book.price / 2
    end

    def visit_fruit(fruit)
        fruit.price / 2
    end
end

class ShoppingCart
    attr_accessor :list, :visitor

    def initialize
        @list = []
        @visitor = nil
    end

    def add(element)
        list << element
    end

    def set_discount_visitor(discount)
        self.visitor = discount
    end

    def accept
        cost = list.sum { |element| element.accept(visitor) }
        puts "Total cost: #{cost}"
    end
end

# Client code
cart = ShoppingCart.new
cart.add(Fruit.new(100, 10, 'Apple'))
cart.add(Book.new(100, 12345))

cart.set_discount_visitor(SundayDiscount.new)
cart.accept

cart.set_discount_visitor(SaleDiscount.new)
cart.accept

=begin 
Total cost: 1000
Total cost: 550
=end